onc-healthit/lantern-back-end
# Adds configuration capabilities using environment variables

Follows the 12 factor app config guidance (https://12factor.net/config) and uses environment variables for configuration parameters.
Uses the viper library to manage the configuration variables.

Not all variables named with "TODO: configuration" were moved into environment variables in the `endpoints/` files as these require additional logic to make that change.

NOTE: the `vendor` directories had A LOT of files added to them. The changes were collapsed into two commits: vendor folder and go.mod related changes (the first commit) and code related changes (the second commit). To make review easier, consider reviewing using the second commit rather than the PR diff.

----
**TESTING**

If you have suggestions on how to _code_ tests for this, please let me know.

**If you haven't setup your DB, follow the instructions here:** https://github.com/onc-healthit/lantern-back-end/pull/8

*ENDPOINT MANAGER*

1. Observe the default values (`endpointmanager/main.go:setupConfig`) and compare those to the values that should work on your machine.
    - [x] Run the tests and ensure the expected result
      1. If they are the same, the tests should **pass**
      1. If they are different, the tests should **fail**
1. Add the environment variables (as described in `README.md`) for the values *that need to be changed* with the correct values.
    - [x] Run the tests and ensure they **pass**
1. Add all environment variables with the correct values.
    - [x] Run the tests and ensure they **pass**
1. Go through each environment variable and modify it to an incorrect value while keeping all the other environment variables as the correct value.
    - [x] Run the tests and ensure they **fail**
      1. Note: for changes to DBHOST, the test won't fail until there's a timeout. Assume that if the test is taking forever that it correctly read in the wrong information and will timeout. Or, if you feel like it, wait a while.
1. Change all environment variables to the correct values
    - [x] Run the tests and ensure they **pass**

*ENDPOINT QUERIER*

1. Observe the default values (`endpoints/main.go:setupConfig`).
1. Start the endpoint querier
    - [x] Check that the `/metrics` endpoint is hosted at the configured port and that logs are being written to the configured log file.
1. Add the environment variables (as described in `README.md`) with *different* values than the default values and restart the endpoint querier.
    - [x] Check that the `/metrics` endpoint is hosted at the port configured in the environment variable and that logs are being written to the log file specified in the environment variable.
----

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-27
  - vendor/ directories have been updated
    - New dependencies have been added
    - Local dependencies have been updated
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
  - vendor/ directories have been updated
    - New dependencies have been added
    - Local dependencies have been updated
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code accomplishes the tasks purpose.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge is complete.
